Fix #12510

XenoBytesX opened the pull request:
### Describe your change:



* [ ] Add an algorithm?
* [✔] Fix a bug or typo in an existing algorithm?
* [✔] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [✔] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [✔] This pull request is all my own work -- I have not plagiarized.
* [✔] I know that pull requests will not be merged if they fail the automated tests.
* [✔] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [✔] All new Python files are placed inside an existing directory.
* [✔] All filenames are in all lowercase characters with no spaces or dashes.
* [✔] All functions and variable names follow Python naming conventions.
* [✔] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [✔] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [✔] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [✔] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #12510 ".

MaximSmolskiy commented:
@XenoBytesX Can you please describe where the error was in the current implementation and why it had to be completely rewritten? I expected that perhaps a minor change to the current implementation would be enough

Additionally, please add a doctest from issue
```
>>> longest_subsequence([28, 26, 12, 23, 35, 39])
[12, 23, 35, 39]
```

MaximSmolskiy reviewed (changes requested):
Please see comment above

XenoBytesX commented:
> @XenoBytesX Can you please describe where the error was in the current implementation and why it had to be completely rewritten? I expected that perhaps a minor change to the current implementation would be enough
> 
> Additionally, please add a doctest from issue
> 
> ```
> >>> longest_subsequence([28, 26, 12, 23, 35, 39])
> [12, 23, 35, 39]
> ```

The current code was failing the testcase mentioned in the issue #12510. The current code was difficult to understand and comparatively slower to what i have implemented.

I have added the doctest and commited it to the pull request.

MaximSmolskiy commented:
I know that current implementation fails some tests (I am author of issue). But I expect that such bugs should be fixed in following way - understand current implementation and fix it (often with fairly minor and simple changes) or write some thoughts why current implementation is very incorrect and cannot be fixed in this way and only then reimplement it

You can add your non-recursive implementation as different separate implementation for solving this problem, but when solving this issue I expect a different approach

P.S. This repository is not about fastest implementations, it is about different implementations of algorithms or to solve problems. So, if we can have two different implementations to solve same problem - it is better to have them both instead of only one - the fastest of them. In this particular case I think that current recursive implementation is also quite important

XenoBytesX commented:
Ok, Understood. I will close this PR and open a new PR for fixing the bug and an another seperate PR with the new implementation. Thanks
